.github/workflows/haskell-ci.yml

```diff
@@ -17,8 +17,23 @@ on:
   - push
   - pull_request
 jobs:
+  tools_changed:
+    continue-on-error: true
+    runs-on: ubuntu-20.04
+    outputs:
+      should_skip: ${{ steps.skip_check.outputs.should_skip }}
+    steps:
+      - id: skip_check
+        uses: fkirc/skip-duplicate-actions@master
+        with:
+          concurrent_skipping: "never"
+          skip_after_successful_duplicate: "true"
+          paths: '["code/**"]'
+          do_not_skip: '["push", "workflow_dispatch", "schedule"]'
   linux:
     name: Haskell-CI - Linux - ${{ matrix.compiler }}
+    needs: tools_changed
+    if: ${{ needs.tools_changed.outputs.should_skip != 'true' }}
```

Comment:
Whatever changes we make in `haskell-ci.yml`, they need to be implemented in `.github/haskell-ci.patch` so that `haskell-ci regenerate` will preserve the modifications/additions.  See https://frasertweedale.github.io/blog-fp/posts/2023-06-04-haskell-ci-use-executable.html#patching-haskell-ci.yml for details.